docs: Expand rule options schema docs (#17198)
eslint/eslint: docs/src/extend/custom-rules.md

Rules may export a `schema` property, which is a [JSON Schema](https://json-schema.org/) format description of a rule's options which will be used by ESLint to validate configuration options and prevent invalid or unexpected inputs before they are passed to the rule in `context.options`.

> If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation.

Reviewer: Just want to call out that some of this behavior will be changing in the next major release v9 due to [RFC-85](https://github.com/eslint/rfcs/tree/main/designs/2021-schema-object-rules) ([draft PR](https://github.com/eslint/eslint/pull/16614)).

>- Requiring rules with options to specify schemas (removing support for rules with options that are missing schemas)
>- Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)

It would ideal if we can take that into account here by:
* Avoid providing recommendations that conflict with the upcoming requirements
* Write the documentation in such a way that it can be easily updated when the new requirements come into effect

Author: > Avoid providing recommendations that conflict with the upcoming requirements

I don't think the docs as written now _recommend_ not providing a schema, it just explains the consequence of not providing one. I can update the line to note that it is explicitly _not_ recommended, and that this option will be removed in v9. Once v9 drops, the change to the docs should be as simple as replacing that paragraph to explain that if you don't provide a `schema` object, your rule will not be permitted to have any options specified (equivalent to `[]`)?

> Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)

I don't think these docs have any conflict with that? They explain `schema`/`meta.schema`, but that change is eliminating the `const rule = function(context, options) { ... }; rule.schema = { ... };` form afaict, which is no longer documented at all?

Author: Updated docs to:

>If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation. This is a legacy behaviour though, and all rules are encouraged to specify a schema to aid developers when configuring your rules. Note that in v9, this behaviour will be deprecated entirely, and not specifying `schema` will force the rule to have no options (equivalent to `[]`) - see the [Require schemas and object-style rules](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) RFC.

Does this cover the bases you wanted to cover?

Author: Also noting in the PR you linked:

> Throw with no-op schema schema: {}

Would it be helpful to add some further rule schema validations to catch obvious cases that will _always_ fail, like a top-level `type: 'object/number/string'`? Could also recurse into `anyOf`/`oneOf`/`allOf` cases and do the same check - i.e. the first `type` to be tested always needs to be an `array`?

Not necessarily suggesting adding those validations in this PR, though I'm happy to take a stab at it separately?

Reviewer: Your documentation tweaks sound reasonable, thanks!

As for catching more bad schemas, I certainly wouldn't mind that, but out of scope for my initial [PR](https://github.com/eslint/eslint/pull/16614) implementing that [RFC](https://github.com/eslint/rfcs/tree/main/designs/2021-schema-object-rules). I think we could consider another breaking change PR with more validation though.

Rules may export a `schema` property, which is a [JSON Schema](https://json-schema.org/) format description of a rule's options which will be used by ESLint to validate configuration options and prevent invalid or unexpected inputs before they are passed to the rule in `context.options`.

> If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation. This is a legacy behaviour though, and all rules are encouraged to specify a schema to aid developers when configuring your rules. Note that in v9, this behaviour will be deprecated entirely, and not specifying `schema` will force the rule to have no options (equivalent to `[]`) - see the [Require schemas and object-style rules](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) RFC.

Reviewer 2: ```suggestion
**Note:** Prior to ESLint v9.0.0, rules without a `schema` are passed their options directly from the config without any validation. In ESLint v9.0.0 and later, rules without schemas will receive an empty array as `context.options`. See the [Require schemas and object-style rules RFC](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) for further details.
```

Author: I'm not sure your suggestion quite covers the proposed strictness in the [RFC](https://github.com/eslint/rfcs/blob/47b9bec8b5d621a70209b29d4a7c237b43d16e33/designs/2021-schema-object-rules/README.md?plain=1#L32C1-L32C1).

I think the way you've phrased it implies that even if a user specifies options for a rule without a schema, they will be ignored and an empty array passed. But the RFC proposes something stricter, i.e. it is an error for a user to specify any options if a rule doesn't provide a schema.

I know the original version doesn't explicitly spell out this consequence either - should we perhaps highlight the consequence of this breaking change for existing schema-less rules? Perhaps the following:

> **Note:** Prior to ESLint v9.0.0, rules without a `schema` are passed their options directly from the config without any validation. In ESLint v9.0.0 and later, rules without schemas will be treated as if they specified a schema of `[]` - this will cause errors for users that specify options for your rule when they upgrade to ESLint v9.0.0+. See the [Require schemas and object-style rules RFC](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) for further details.

Author: Yep, even more short and sweet. Will merge in now - I'm just making these changes directly on the branch rather than committing the suggestions in the GH UI 😄

Reviewer 2: That's fine, just make sure the conversations are marked as "Resolved" so we know what's been done.

Reviewer: Sorry late reply but it looks good.

> When validating a rule's options, there are 3 main preparation/validation steps:

Reviewer 2: ```suggestion
When validating a rule's options, there are three steps:
```

> 1. If the options are not an array, the value is wrapped into an array (e.g. `"off"` becomes `["off"]`)
> 2. ESLint will validate the first element of the options array as a severity (`"off"`, `"warn"`, `"error"`, `0`, `1`, `2`)
> 3. If the rule is enabled, then the remaining elements of the array are sliced and the rule's schema validation is run on this sliced options array (e.g. `["warn", "never", { someOption: 5 }]` results in the schema being applied to `["never", { someOption: 5 }]`)

Reviewer 2: You are mixing up terms here. "options" means anything passed in the config except the severity, so we need to clean this up a bit.

```suggestion
1. If the rule config is not an array, then the value is wrapped into an array (e.g. `"off"` becomes `["off"]`); if the rule config is an array then it is used directly.
2. ESLint validates the first element of the rule config array as a severity (`"off"`, `"warn"`, `"error"`, `0`, `1`, `2`).
3. If the rule is enabled, then the other elements of the array (everything except the severity) are copied an options array (into the `context.options` property) and the rule's schema validation is run on this array (e.g. `["warn", "never", { someOption: 5 }]` results in the schema being applied to `["never", { someOption: 5 }]`).
```

> Note that this means that the rule schema cannot see or validate the severity, it only sees the array elements _after_ the severity.

Reviewer 2: ```suggestion
Note: this means the rule schema cannot validate the severity., The rule schema only validates the array elements _after_ the severity in a rule config.
```

Author: Nice. I'm also going to add an extra sentence on the end to really spell it out:

> There is no way for a rule to know what severity it is configured at.

There are two formats for a rule's exported `schema`:

* An array of JSON Schema objects

>     * Each element will be checked against the same position in the sliced options array
>     * If the sliced options array has fewer elements than there are schemas, then the unmatched schemas are ignored
>     * If the sliced options array has more elements than there are schemas, the validation fails

Reviewer 2: "sliced options" won't make sense to most folks.

```suggestion
    * Each element will be checked against the same position in the `context.options` array.
    * If the `context.options` array has fewer elements than there are schemas, then the unmatched schemas are ignored.
    * If the `context.options` array has more elements than there are schemas, then the validation fails.
```

    * Two important consequences:
        * Empty rule options beyond severity are always valid (e.g. `["error"]` will be a valid rule option)

>         * An empty array _enforces no options_ for the rule

Reviewer 2: I've re-read this a few times and I'm still not clear on what these lines are trying to say. 

Consequences of what?

Is the first bullet saying that you are not required to pass options beyond severity?

For the second bullet, what is enforcing no options? An empty array where?

Author: Sure, I've tried to clarify the wording on this to highlight the always valid/always error cases a bit more:

>    * There are two important consequences to using this format:
>        * It is _always valid_ for a user to provide no options to your rule (beyond severity)
>        * If you specify an empty array, then it is _always an error_ for a user to provide any options to your rule (beyond severity)

> * A full JSON Schema object that will validate the sliced options array

Reviewer 2: ```suggestion
* A full JSON Schema object that will validate the `context.options` array
```

>     * Be very careful that you are _always_ validating against an array, even if you only expect one options value/object to be supplied

Reviewer 2: ```suggestion
    * The schema should assume an array of options to validate even if your rule only accepts one option.
```

module.exports = {
    meta: {
        schema: [
            {
                enum: ["always", "never"]
            },
            {
                type: "object",
                properties: {
                    exceptRange: { type: "boolean" }
                },
                additionalProperties: false
            }
        ]
    }
};
```

And here is the equivalent object-based schema:

```js
// Valid options:
// "yoda": "warn"
// "yoda": ["error"]
// "yoda": ["error", "always"]
// "yoda": ["error", "never", { "exceptRange": true }]
// Invalid options:
// "yoda": ["warn", "never", { "exceptRange": true }, 5]
// "yoda": ["error", { "exceptRange": true }, "never"]
module.exports = {
    meta: {
        schema: {
            type: "array",
            minItems: 0,
            maxItems: 2,
            items: [
                {
                    enum: ["always", "never"]
                },
                {
                    type: "object",
                    properties: {
                        exceptRange: { type: "boolean" }
                    },
                    additionalProperties: false
                }
            ]
        }
    }
};
```

> Object schemas are more powerful though, and can be much more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specifed (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.

Reviewer 2: ```suggestion
Object schemas can be more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specified (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.
```

Author: I've also updated the example code to note use `yoda` as the rule name, to avoid any confusion with the real rule.

// Invalid options:
// "yoda": "warn"
// "yoda": ["error"]
// "yoda": ["warn", 15]
// "yoda": ["warn", 7, { }]
// "yoda": ["error", 3, "on"]
// "yoda": ["warn", 7, { someOtherProperty: 5 }]
// "yoda": ["warn", 7, { someNonOptionalProperty: false, someOtherProperty: 5 }]
module.exports = {
    meta: {
        schema: {
            type: "array",
            minItems: 1, // Can't specify only severity!
            maxItems: 2,
            items: [
                {
                    type: "number",
                    minimum: 0,
                    maximum: 10
                },
                {
                    anyOf: [
                        {
                            type: "object",
                            properties: {
                                someNonOptionalProperty: { type: "boolean" }
                            },
                            required: ["someNonOptionalProperty"],
                            additionalProperties: false
                        },
                        {
                            enum: ["off", "strict"]
                        }
                    ]
                }
            ]
        }
    }
}
```

> And note the easy pitfall of specifying a schema that will _always_ fail because it doesn't account for the array being validated:

Reviewer 2: ```suggestion
Remember, rule options are always an array, so be careful not to specify a schema for a non-array type at the top level. Here's an example schema that will always fail validation:
And note the easy pitfall of specifying a schema that will always fail because it doesn't account for the array being validated:
```

Author: I've also added another sentence to really spell out the consequence of doing this wrong:

> If your schema does not specify an array at the top-level, users can _never_ enable your rule, as their configuration will always be invalid when the rule is enabled.

```js
// Possibly trying to validate ["error", { someOptionalProperty: true }], but:

> // THERE ARE NO POSSIBLE VALID OPTIONS FOR THIS SCHEMA WHEN THE RULE IS ENABLED

Reviewer 2: ```suggestion
// When the rule is enabled, this options will always fail validation because the options are an array which doesn't match "object"
```

};
```

> And as an example for the full JSON Schema object form:

Reviewer 3: This sounds like the following will be the same schema, just in the JSON Schema object form, but it isn't the same. Perhaps it would be better to really show how the schema for `yoda`  rule options would look like in the JSON Schema object form?

Author: Sure, I was trying to show that there's more/different stuff you can do with the object schema, but I can see how it looks confusing.

I can split it into two examples, one explicitly showing the array-equivalent object schema, and a second showing off how the object schema can diverge in behaviour?
